scripts/linters/pylint_extensions.py

```diff
@@ -385,6 +407,104 @@ def check_functiondef_params(self, node, node_doc):
+            None if (arg.name in self.not_needed_param_in_docstring)
+            else (arg.name + ':') for arg in arguments_node.args)
+        currently_in_args_section = False
+        # When we are in the args section and a line ends in a colon,
+        # we can ignore the indentation styling in the next section of
+        # in_description, hence a freeform section.
+        currently_in_freeform_section = False
+        args_indentation = 0
+        if node.doc:
+            # Process the docstring line by line.
+            for line in node.doc.splitlines():
+                line_with_no_leading_whitespaces = line.lstrip()
+                current_line_indentation = (
+                    len(line) - len(line_with_no_leading_whitespaces))
+                parameter = re.search(
+                    '^[^:]+:',
+                    line_with_no_leading_whitespaces)
+                # Check for empty lines and ignore them.
+                if len(line.strip()) == 0:
+                    continue
+                # Check for the start of an Args: section and check the correct
+                # indentation.
+                if line_with_no_leading_whitespaces.startswith('Args:'):
+                    args_indentation = current_line_indentation
+                    # The current args indentation is incorrect.
+                    if current_line_indentation % 4 != 0:
+                        self.add_message(
+                            'incorrect-indentation-for-arg-header-doc',
+                            node=node)
+                        # Since other checks are based on relative indentation,
+                        # we need to fix this indentation first.
+                        break
+                    currently_in_args_section = True
+                # Check for parameter section header by checking that the
+                # parameter is in the function arguments set. We also check for
+                # arguments that start with * which means it's autofill and will
+                # not appear in the node args list so we handle those too.
+                elif (currently_in_args_section and parameter
+                      and ((parameter.group(0).strip('*')
+                            in expected_argument_names) or
+                           re.search(
+                               br'\*[^ ]+: ',
+                               line_with_no_leading_whitespaces))):
+                    words_in_line = line_with_no_leading_whitespaces.split(' ')
+                    currently_in_freeform_section = False
+                    # Check if the current parameter section indentation is
+                    # correct.
+                    if current_line_indentation != (
+                            args_indentation + 4):
+                        # Use the first word in the line to identify the error.
+                        beginning_of_line = (
+                            None if
+                            not words_in_line
```

Comment:
Optional: Might be slightly nicer to do "words_in_line[0] if words_in_line else None" (i.e. handle the 'standard' case first). Ditto below.